mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir

```diff
@@ -187,3 +187,49 @@ module attributes {transform.with_named_sequence} {
     transform.yield
   }
 }
+
+// -----
+
+
+//       CHECK:   #[[MAP:.*]] = affine_map<(d0, d1, d2, d3) -> (d1, 0, d3)>
+//       CHECK:   func.func @transfer_read_reduce_rank_scalable(
+//  CHECK-SAME:       %[[ARG_0:.*]]: memref<?x?x?x?xf32>) -> vector<8x[4]x2x3xf32> {
+//       CHECK:     %[[C0:.*]] = arith.constant 0 : index
+//       CHECK:     %[[TFR:.*]] = vector.transfer_read %arg0[%[[C0]], %[[C0]], %[[C0]], %[[C0]]]{{.*}} permutation_map = #[[MAP]]} : memref<?x?x?x?xf32>, vector<[4]x2x3xf32>
+//       CHECK:     %[[BC:.*]] = vector.broadcast %[[TFR]] : vector<[4]x2x3xf32> to vector<8x[4]x2x3xf32>
+//       CHECK:     return %[[BC]] : vector<8x[4]x2x3xf32>
+func.func @transfer_read_reduce_rank_scalable(%mem: memref<?x?x?x?xf32>) -> vector<8x[4]x2x3xf32> {
```

Comment:
How is this test different to `@permutation_with_mask_xfer_read_scalable`? Both seem to check for `vector.transfer_read` from `memref` into a scalable vector?

Reply:
There is a transposition in the permutation map of `permutation_with_mask_xfer_read_scalable`. It would hence also trigger `TransferReadPermutationLowering` which IIRC would fail without https://github.com/llvm/llvm-project/pull/91987 . Changes were not completely independent, this one is a bit more _unit_ test-like

Reply:
> There is a transposition in the permutation map of permutation_with_mask_xfer_read_scalable.

Cool, right now that's very unclear, unless you happen to know what to look for. Such distinction should be captured either:
* through comments, and/or
* test function name.

(preferably both). That's currently not the case, so lets fix that.  I'm suggesting the following plan-of-action:
* add a big bold comment to document what's being tested here, something similar to what's on L94-L101
* add `no_transpose` to the test functions names that you added.

I can take it from there.

I have two more high-level points. A bit tangential to this PR - more like TODOs for myself 😅 

### What test should be included in vector-transfer-permutation-lowering.mlir?

Btw, based on the file name and the patterns being tested here:
* [vector-transfer-permutation-lowering.mlir](https://github.com/llvm/llvm-project/pull/92426/files/12f68ff1983dcbd3e3ae5effb916f853f512600a#diff-87b7cedb5d39b7a7598c7dcf29c95ee7cf6522f85163c3044d23e19f1e6f232c)
* `transform.apply_patterns.vector.transfer_permutation_patterns`,

there should be no non-permutation examples in this file, right? To make things even more confusing, these patterns are also tested in:
* transform-vector.mlir
*  vector-transfer-to-vector-load-store.mlir

😱 :) I'm on a mission to tidy this up a bit (while adding tests for scalable vectors).

### Note on populateVectorTransferPermutationMapLoweringPatterns
Separately, it's bit confusing that `TransferOpReduceRank` is added to `populateVectorTransferPermutationMapLoweringPatterns`:
https://github.com/llvm/llvm-project/blob/77db8b08c8b186c2625f8dfb26bb976561b43c4c/mlir/lib/Dialect/Vector/Transforms/LowerVectorTransfer.cpp#L410-L416

I think that it would be easier to "categorise" tests if the granularity was lower. Another PR though. I can handle this.

Reply:
Makes sense to me, it deserves refactoring and standardization. I could have done better on commenting and naming.

> add a big bold comment to document what's being tested here, something similar to what's on L94-L101

Alternatively, separating tests with a big bold comment by pattern they are actually testing from the set could make sense.

> add no_transpose to the test functions names that you added.

I think this quite goes against the idea of having a permutation specific file. as you mention there :

> there should be no non-permutation examples in this file, right? To make things even more confusing, these patterns are also tested in:

But I think we should rather emphasize on the fact it is not a permutation lowering but a lowering based on the permutation **_map_** . Hence the `no_transpose` answer works.

>  these patterns are also tested in:

I would not mind about `transform-vector.mlir` as it is a full pipeline test. The goal is to show and test a complete lowering which makes sense. A bit like the e-2-e examples you added for SME.
For vector-transfer-to-vector-load-store.mlir , I think this file deserves improvements too. (such as the repetition of same transform sequences. ) The concept is to combine it with lower_transfer to test the complete lowering of xfer ops. it is not very _**unit**_ test like but it shows how to properly lower xfer ops which is a benefit. Mixed feelings. I think there should not be `transfer_permutation_patterns` specific tests here, but I find sensible to test this combination of patterns. We should maybe ask @ftynse .

> it's bit confusing that TransferOpReduceRank is added to populateVectorTransferPermutationMapLoweringPatterns:

Indeed, but after all it is a lowering based on the permutation map of a TransferRead. Hence, I think than rather removing `TransferOpReduceRank`,  the ideal solution is to rename `TransferOpReduceRank`  to make clear it matches patterns in the permutation map or renaming `populateVectorTransferPermutationMapLoweringPatterns` to reflect it is not mandatory a permutation. This set of patterns is useful as a whole for the community as a lowering pass.

I would also add something. As @hanhanW  mentioned in https://github.com/llvm/llvm-project/pull/93664#pullrequestreview-2089507138  we could get rid of  `/// ``` and standardize it.
` in Doxygen descriptions for improved readability.

Reply:
> in Doxygen descriptions for improved readability.

It is also helpful when I parse the comment in the code directly. I'd appreciate if someone can help make it consistent. :)

Reply:
The syntax is mostly for markdown to me, so I don't know why people started adding such style to comments.

Reply:
> But I think we should rather emphasize on the fact it is not a permutation lowering but a lowering based on the permutation map . Hence the no_transpose answer works.

That's spot-on and something that I missed. In fact, this is documented here:
https://github.com/llvm/llvm-project/blob/77db8b08c8b186c2625f8dfb26bb976561b43c4c/mlir/include/mlir/Dialect/Vector/Transforms/LoweringPatterns.h#L175-L228

With this in mind, please ignore my comments under this heading:
> Note on populateVectorTransferPermutationMapLoweringPatterns

In fact, sounds like [vector-transfer-permutation-lowering.mlir](https://github.com/llvm/llvm-project/pull/92426/files/12f68ff1983dcbd3e3ae5effb916f853f512600a#diff-87b7cedb5d39b7a7598c7dcf29c95ee7cf6522f85163c3044d23e19f1e6f232c) is the right file for all patterns under `populateVectorTransferPermutationMapLoweringPatterns`.

Going back to the problem at hand ... This

> Makes sense to me, it deserves refactoring and standardization.

and this:

> Alternatively, separating tests with a big bold comment by pattern they are actually testing from the set could make sense.

First step in this direction: https://github.com/llvm/llvm-project/pull/95529. Let me know if that makes sense and I will prepare more tests.

> I could have done better on commenting and naming.

Not your fault, it's hard to navigate this ATM (this is also why it takes me ages to review things). Clearly our test coverage is not great either. In general, Vector needs some TLC 😅 I really appreciate you helping us here 🙏🏻

Reply:
> First step in this direction: https://github.com/llvm/llvm-project/pull/95529. Let me know if that makes sense and I will prepare more tests.

Next one: https://github.com/llvm/llvm-project/pull/96033